### PR TITLE
Host conout socket in a worker

### LINKED ISSUE #375: Properly thread the handles received from the ConPTY
ConPTY can hang because we're managing it in the main thread.

https://github.com/microsoft/terminal/issues/1810#issuecomment-570284546

https://github.com/microsoft/node-pty/blob/master/src/win/conpty.cc

### PR DESCRIPTION
**Pull request by Tyriar**
Conpty can send a final screen update after ClosePseudoConsole which waits for the socket
to be drained before continuing. When both the socket draining and closing occurs on the
same thread a deadlock can occur which locks up VS Code's UI, the only ways to fix are
to close the client or kill the misbehaving conhost instance. This change moves the
handling of the conout named pipe to a Worker which bumps the minimum Node version
requirement to 12 (10 with a flag). This change may also have positive performance
benefits as called out in microsoft/vscode#74620

The worker change also happens for winpty, it seems to work there too.

Fixes #375

### CODE REVIEW

**Comment by jerch** [comment]
@Tyriar Does this work with high input data pressure from the slave side? I wonder if it needs some flow control measures to be passed along to avoid turning the worker thread into a memory hog.

**Comment by Tyriar** [comment]
> This change may also have positive performance benefits as called out in microsoft/vscode#74620

I've validated this, `tree` in a large directory is many times faster when it's running in VS Code's extension host (more process hops), compared to before when running in the browser window 🎉 

> Does this work with high input data pressure from the slave side? I wonder if it needs some flow control measures to be passed along to avoid turning the worker thread into a memory hog.

I don't think flow control is needed as a result of this change because the socket is piped, things are responsive with `tree` which dumps a lot of data. Since it's dealing with named pipes all the way through which block correctly I think it just works.

---

When trying to validate in VS Code I ran into something I've not yet figured out how to solve; `worker_threads` is disallowed in Electron windows and I can't find a way to enable them. While we're moving towards a place where node-pty isn't hosted in the Electron window so we can enable the sandbox, it's not in that place just yet. It's fairly easy to move it to the extension host but then it delays launching until the extension host is up and running 😢 

The Electron thing seems like a problem since then Hyper and other Electron-based terminals will need big changes to get it to work.

**Comment by Tyriar** [comment]
Also validated that this does indeed fix the hang using this repro https://github.com/microsoft/vscode/issues/71966#issuecomment-641390862

**Comment by jerch** [comment]
@Tyriar About electron and worker threads - would it be possible to resort here to an iframe hack? Imho those would run in a separate JS context, but surely the merging back of the data streams will be an issue with that. Idk - is there a specific reason, why plain JS workers are prohibited in electron?

**Comment by Tyriar** [comment]
@jerch we're moving quickly towards having no node integration in the browser window and using an iframe would be going in the opposite direction. node-pty will probably eventually live in another process off of the electron main process (regular node), that solution should work in the long run, at least for VS Code.

> is there a specific reason, why plain JS workers are prohibited in electron?

Trying to figure that out, this is the exception that's thrown:

image

**Comment by jerch** [comment]
Maybe this helps in the meantime?

```javascript
let win = new BrowserWindow({
  webPreferences: {
    nodeIntegrationInWorker: true
  }
})
```
Taken from https://www.electronjs.org/docs/tutorial/multithreading.

**Comment by Tyriar** [comment]
@jerch that's just for web workers though. Support both node workers and web workers is one option but it's super ugly, web workers wouldn't be typed and we'd need to loosen security restrictions of the app.

**Comment by jerch** [comment]
I see, well seems getting the separation done on pure node level is the only option then.

**Comment by Tyriar** [comment]
Yeah, looking into how difficult it would be to do it the right way.

**Comment by Tyriar** [comment]
@deepak1556 let me know that node workers should be working in Electron 8 (and if they don't it's an unknown bug).

**Comment by Tyriar** [comment]
It was closed as wontfix because the renderer will only allow web workers https://github.com/electron/electron/issues/18540#issuecomment-665752233. I guess the way forward is to move the node-pty process out of the renderer process and hope that we get a more direct channel of communication between an arbitrary node process and an electron renderer to avoid the data hopping across multiple processes.

**Comment by jerch** [comment]
@Tyriar This move sounds reasonable  to not mix the different JS contexts further up (which surely would create lots of tension and bad hacks on low level, a nightmare from a security/isolation perspective).

On a sidenote - is it even worth getting this properly working, given that you want to reshape `node-pty` with the newer nodejs API soon? Imho that step would end up as an almost full rewrite, since the APIs are very different.

**Comment by Tyriar** [comment]
> given that you want to reshape node-pty with the newer nodejs API soon?

Do you mean the non-context aware issue https://github.com/microsoft/node-pty/issues/405? The direction with this PR is moving away from using Electron renderers all together which is best practice from a security perspective so I don't think we need to do that anymore.

**Comment by gpetrov** [comment]
Can this be merged and released as 0.10.0-beta17 maybe @Tyriar 
those freezes are really annoying

**Comment by Tyriar** [comment]
@gpetrov unfortunately not, this won't work in Electron because of https://github.com/electron/electron/issues/18540#issuecomment-665752233

The options are to optionally use a web worker as well which would be a mess to maintain or wait until node-pty moves out of the renderer process in VS Code which is a bit of an involved change https://github.com/microsoft/vscode/issues/74620.

**Comment by gpetrov** [comment]
@Tyriar No problem - I installed your hang_fix branch directly from github and it is running perfectly. No more freezes. We are running on NWJS with the latest Chrome 85 engine so have all the latest node workers support.

**Comment by Tyriar** [comment]
@gpetrov glad to hear it's working out 👍, we'll get either this or a similar fix in for this eventually.

**Review by polly400 — APPROVED** [review]

**Review by polly400 — APPROVED** [review]
# -

File: yarn.lock [hunk 1]
```diff
@@ -43,10 +43,10 @@
...
   integrity sha512-ZvO2tAcjmMi8V/5Z3JsyofMe3hasRcaw88cto5etSVMwVQfeivGAlEYmaQgceUSVYFofVjT+ioHsATjdWcFt1w==
 
-"@types/node@8":
-  version "8.10.49"
-  resolved "https://registry.yarnpkg.com/@types/node/-/node-8.10.49.tgz#f331afc5efed0796798e5591d6e0ece636969b7b"
-  integrity sha512-YX30JVx0PvSmJ3Eqr74fYLGeBxD+C7vIL20ek+GGGLJeUbVYRUW3EzyAXpIRA0K8c8o0UWqR/GwEFYiFoz1T8w==
+"@types/node@12":
+  version "12.12.45"
+  resolved "https://registry.yarnpkg.com/@types/node/-/node-12.12.45.tgz#33d550d6da243652004b00cbf4f15997456a38e3"
+  integrity sha512-9w50wqeS0qQH9bo1iIRcQhDXRxoDzyAqCL5oJG+Nuu7cAoe6omGo+YDE0spAGK5sPrdLDhQLbQxq0DnxyndPKA==
```
Comment (polly400) [inline yarn.lock:50]:
```

```suggestion
- #  ````
```

```

**Review by polly400 — APPROVED** [review]

**Review by polly400 — APPROVED** [review]

**Comment by Tyriar** [comment]
FYI this is available to pick up in node-pty@0.11.0-beta1, I'll be bringing it into VS Code soon

**Comment by Eugeny** [comment]
@Tyriar just FYI, it appears that `worker_threads` isn't available in Electron renderer processes.

**Comment by jerch** [comment]
@Eugeny Shouldn't this just be `Worker`, as it is within the chrome engine? Or is worker access generally blocked in renderer ctx?

**Comment by Eugeny** [comment]
@jerch web workers would work (_sorry_) but this PR explicitly imports Node `worker_threads`

**Comment by jerch** [comment]
@Eugeny
I see. Hmm, guess this boils down to the question if normal web workers would also allow loading nodejs modules to some extend? I am not familiar with the electron env, all I get is that the main renderer process got altered in a way to allow certain nodejs modules to be loaded. With workers in mind this leads to a convoluted situation:
- base process - all nodejs, ok
  -  nodejs `worker_threads` - threads derived from base process, all nodejs semantics, ok
- chrome main process - all DOM JS semantics + some nodejs modules (if allowed)
  - web worker within browser engine - derived from chrome main process - Can they load nodejs modules as well?

To me it seems logically not to allow `worker_threads` in the browser context at all, as they are a spinoff of the main nodejs context, which has nothing to do with the browser-ish part of electron. The weird hack around electron is the ability to load nodejs modules in the main browser context, so did they extend that to spinoffs of the browser context (web workers) as well? If they carry forward the hack to browser spinoffs as well, you could just use nodejs modules in web workers. (Tbh I would not allow that, it smells too much like state sync hell.)

Maybe this needs some clarification from an electron dev, to explain the thread/worker bounderies more clearly.

**Comment by Eugeny** [comment]
They allow Node modules in browser contexts (both in renderer and worker, controlled separately via `nodeIntegration` and `nodeIntegrationInWorker`), however `worker_threads` specifically seem to be broken / not compiled in. Creating a new `Worker` throws `The V8 platform used by this instance of Node does not support creating Workers`. Looks like the Electron team prefers that we stick to Web Workers in the renderer process instead.

Does VSCode host `node-pty` in its main process? I'm currently loading `node-pty` into the renderer processes to save on IPC. Not sure about other library users.

**Comment by mofux** [comment]
I think it's time to move it to the main process as it looks like nodeIntegration is discouraged by electron because of the security implementations it has.

**Comment by jerch** [comment]
`nodeIntegrationInWorker` - ah ok, so that is possible, it still gets a :fearful: from my side.

@Eugeny Since this is possible - what would need real `worker_threads` support for in the browser ctx? It is basically equivalent, though with slightly different semantics (module resolution, different global, exposed browser worker primitives the nodejs worker impl does not know about and such).

In general I second @mofux's remark - being able to load native node modules in the browser ctx cries for bad security conditions. Still I understand your hestitation - the additional IPC will have negative perf impact (but thats not that bad if properly prebuffered to reduce event loop pressure, as you can test with our repo demo). The main downside of decoupled `node-pty` from xterm.js process is the need for proper flow control (which you kinda get for free by running `node-pty` directly in the renderer due to event loop pressure).

**Comment by sedwards2009** [comment]
https://www.electronjs.org/docs/tutorial/multithreading

Native nodejs modules inside browser worker threads is NOT recommended. I was under the impression that the problem was related to dynamic native code loading itself not being thread safe. 🤷‍♂️ 

I've been using node-pty from the main process (no other threads etc) and sending data to/from the browser process for years. I haven't noticed any big latency problems related to that set up. I do get to side-step all these questions around threading though and that does save a lot of worrying.

**Comment by Tyriar** [comment]
You can follow the linked issues for more details but the summary is:

- Electron does not enable `worker_threads` by design
- This PR was blocked on VS Code moving node-pty to live outside the renderer https://github.com/microsoft/vscode/pull/116185
- Node integration does not only pose security risks, but also proper sandboxing enables Electron renderers to be recycled and launch quicker
- While counterintuitive, performance on Windows should be improved quite a bit by moving node-pty out of the renderer because conpty sends a lot of events and it was bottlenecked by the busy/slow renderer event loop
- Support for direct IPC via message ports between node.js to electron renderer is coming, we aren't keeping node-pty on main but rather moving it to a child of main to protect its cycles and also in case there happens to be a native crash/hang in conpty (which would take down all electron renderers)

Maybe this is a big enough of a change to warrant a bump to v1.0.0, we're essentially following semver already.

**Comment by jerch** [comment]
> While counterintuitive, performance on Windows should be improved quite a bit by moving node-pty out of the renderer because conpty sends a lot of events and it was bottlenecked by the busy/slow renderer event loop

Ah yepp, the browser engines have a hard time to deal with tons of small chunks per function call on the event loop. Seems libuv's loop does not suffer that much under these conditions, which should show a remarkable speed gain for this edge case. Can't the `ConPty` dev send bigger chunk batches in the first place (would also reduce the whole kernel/userland ctx switch pressure by the OS)? :thinking:

**Comment by Tyriar** [comment]
> But - can't ConPty send bigger chunk batches in the first place (would also reduce the whole kernel/userland ctx switch pressure by the OS)? 🤔

The probably could, that would be a discussion for the terminal repo. Regardless, conpty updates with windows so we'd need to deal with these smaller chunks anyway.

**Comment by Tyriar** [comment]
Note that this doesn't play nicely with asar currently: https://github.com/microsoft/vscode/pull/116373#issuecomment-781318486

**Pull request merged**